In [14]:
import numpy as np
from keras.datasets import mnist
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers import Dense
from keras.layers.recurrent import SimpleRNN, LSTM, GRU
from keras.optimizers import Adam

### 使用RNN前的定義

In [2]:
# 將28*28的圖像看作是序列(共28序列，每行序列有28個(長度))
# 數據長度:一行有28個像素
input_size = 28
# 序列長度:一共有28行
time_steps = 28
# 隱藏層cell個數(RNN藉由每個RNN的cell連結成一長鏈)
cell_size = 50

### 1_載入數據

In [3]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [4]:
print('x_shape: ', x_train.shape)
print('y_shape: ', y_train.shape)

x_shape:  (60000, 28, 28)
y_shape:  (60000,)


### 2_數據集處理

In [5]:
# (60000, 28, 28)剛好符合RNN的數據型態需求 => (data筆數, 28行序列(time_steps), 每行序列28個個數(input_size)
x_Train = x_train / 255.0   
x_Test = x_test / 255.0 

In [6]:
# 對label轉成one-hot
y_Train_onehot = np_utils.to_categorical(y_train, num_classes=10)
y_Test_onehot = np_utils.to_categorical(y_test, num_classes=10)

### 3-1_創建模型

In [7]:
# 創建順序模型
model = Sequential()

# 循環神經網路(RNN, Recurrent Neural Network)
model.add(SimpleRNN(
    units=cell_size, # 輸出層(這裡指隱藏層個數)，共50個cell
    input_shape=(time_steps, input_size)  # 輸入
))

# 輸出層
model.add(Dense(10, activation='softmax'))

In [8]:
# 定義優化器，loss function, 並在訓練過程中計算準確率
adam = Adam(learning_rate=1e-4)

# 編譯
model.compile(
    optimizer = adam,
    loss = 'categorical_crossentropy',   # 此處只改loss function為交叉熵
    metrics = ['accuracy']
)

### 3-2 訓練模型

In [9]:
model.fit(x_Train, y_Train_onehot, batch_size=64, epochs=10)

Epoch 1/10
60000/60000 [==============================] - 7s 122us/step - loss: 1.5197 - accuracy: 0.5003
Epoch 2/10
60000/60000 [==============================] - 8s 136us/step - loss: 0.8970 - accuracy: 0.7312
Epoch 3/10
60000/60000 [==============================] - 7s 124us/step - loss: 0.6848 - accuracy: 0.7988
Epoch 4/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.5672 - accuracy: 0.8364
Epoch 5/10
60000/60000 [==============================] - 8s 127us/step - loss: 0.4894 - accuracy: 0.8605
Epoch 6/10
60000/60000 [==============================] - 7s 120us/step - loss: 0.4363 - accuracy: 0.8757
Epoch 7/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.3977 - accuracy: 0.8874
Epoch 8/10
60000/60000 [==============================] - 7s 125us/step - loss: 0.3687 - accuracy: 0.8953
Epoch 9/10
60000/60000 [==============================] - 7s 123us/step - loss: 0.3457 - accuracy: 0.9014
Epoch 10/10
60000/60000 [=====================

### 3-3_評估模型

In [13]:
loss, accuracy = model.evaluate(x_Train, y_Train_onehot)
print('\ntrain loss: ', loss)
print('train accuracy: ', accuracy)
print('\n')
loss, accuracy = model.evaluate(x_Test, y_Test_onehot)
print('\ntest loss: ', loss)
print('test accuracy: ', accuracy)

60000/60000 [==============================] - 4s 75us/step

train loss:  0.323143792450428
train accuracy:  0.9077000021934509


10000/10000 [==============================] - 1s 76us/step

test loss:  0.3298914443016052
test accuracy:  0.9061999917030334


### 小結：
* Simple RNN使用在手寫辨識上面(在不調整任何參數情況下)準確率90.6%，算普通。

----------------------

## 試看看其他RNN模型1- GRU

In [16]:
# 將28*28的圖像看作是序列(共28序列，每行序列有28個(長度))
# 數據長度:一行有28個像素
input_size = 28
# 序列長度:一共有28行
time_steps = 28
# 隱藏層cell個數(RNN藉由每個RNN的cell連結成一長鏈)
cell_size = 50

# 載入數據集
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# 調整資料型態
# 歸一化
x_Train = x_train / 255.0   
x_Test = x_test / 255.0 
# 對label轉成one-hot
y_Train_onehot = np_utils.to_categorical(y_train, num_classes=10)
y_Test_onehot = np_utils.to_categorical(y_test, num_classes=10)

# 創建順序模型
model_GRU = Sequential()
# 循環神經網路(RNN, Recurrent Neural Network)
model_GRU.add(GRU(
    units=cell_size, # 輸出層(這裡指隱藏層個數)，共50個cell
    input_shape=(time_steps, input_size)  # 輸入
))
# 輸出層
model_GRU.add(Dense(10, activation='softmax'))

# 定義優化器，loss function, 並在訓練過程中計算準確率
adam = Adam(learning_rate=1e-4)

# 編譯
model_GRU.compile(
    optimizer = adam,
    loss = 'categorical_crossentropy',   # 此處只改loss function為交叉熵
    metrics = ['accuracy']
)

# 訓練模型
model_GRU.fit(x_Train, y_Train_onehot, batch_size=64, epochs=10)

# 評估模型
loss, accuracy = model_GRU.evaluate(x_Train, y_Train_onehot)
print('\ntrain loss: ', loss)
print('train accuracy: ', accuracy)
print('\n')
loss, accuracy = model_GRU.evaluate(x_Test, y_Test_onehot)
print('\ntest loss: ', loss)
print('test accuracy: ', accuracy)

Epoch 1/10
60000/60000 [==============================] - 17s 279us/step - loss: 1.8386 - accuracy: 0.3730
Epoch 2/10
60000/60000 [==============================] - 19s 315us/step - loss: 0.9576 - accuracy: 0.6925
Epoch 3/10
60000/60000 [==============================] - 18s 306us/step - loss: 0.7396 - accuracy: 0.7621
Epoch 4/10
60000/60000 [==============================] - 19s 319us/step - loss: 0.5741 - accuracy: 0.8250
Epoch 5/10
60000/60000 [==============================] - 18s 300us/step - loss: 0.4616 - accuracy: 0.8645
Epoch 6/10
60000/60000 [==============================] - 18s 306us/step - loss: 0.3864 - accuracy: 0.8893
Epoch 7/10
60000/60000 [==============================] - 20s 329us/step - loss: 0.3356 - accuracy: 0.9046
Epoch 8/10
60000/60000 [==============================] - 20s 326us/step - loss: 0.3000 - accuracy: 0.9143
Epoch 9/10
60000/60000 [==============================] - 18s 295us/step - loss: 0.2731 - accuracy: 0.9212
Epoch 10/10
60000/60000 [============

-----------

## 試看看其他RNN模型2- LSTM

In [17]:
# 將28*28的圖像看作是序列(共28序列，每行序列有28個(長度))
# 數據長度:一行有28個像素
input_size = 28
# 序列長度:一共有28行
time_steps = 28
# 隱藏層cell個數(RNN藉由每個RNN的cell連結成一長鏈)
cell_size = 50

# 載入數據集
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# 調整資料型態
# 歸一化
x_Train = x_train / 255.0   
x_Test = x_test / 255.0 
# 對label轉成one-hot
y_Train_onehot = np_utils.to_categorical(y_train, num_classes=10)
y_Test_onehot = np_utils.to_categorical(y_test, num_classes=10)

# 創建順序模型
model_LSTM = Sequential()
# 循環神經網路(RNN, Recurrent Neural Network)
model_LSTM.add(LSTM(
    units=cell_size, # 輸出層(這裡指隱藏層個數)，共50個cell
    input_shape=(time_steps, input_size)  # 輸入
))
# 輸出層
model_LSTM.add(Dense(10, activation='softmax'))

# 定義優化器，loss function, 並在訓練過程中計算準確率
adam = Adam(learning_rate=1e-4)

# 編譯
model_LSTM.compile(
    optimizer = adam,
    loss = 'categorical_crossentropy',   # 此處只改loss function為交叉熵
    metrics = ['accuracy']
)

# 訓練模型
model_LSTM.fit(x_Train, y_Train_onehot, batch_size=64, epochs=10)

# 評估模型
loss, accuracy = model_LSTM.evaluate(x_Train, y_Train_onehot)
print('\ntrain loss: ', loss)
print('train accuracy: ', accuracy)
print('\n')
loss, accuracy = model_LSTM.evaluate(x_Test, y_Test_onehot)
print('\ntest loss: ', loss)
print('test accuracy: ', accuracy)

Epoch 1/10
60000/60000 [==============================] - 17s 276us/step - loss: 1.5537 - accuracy: 0.4921
Epoch 2/10
60000/60000 [==============================] - 17s 287us/step - loss: 0.7236 - accuracy: 0.7847
Epoch 3/10
60000/60000 [==============================] - 17s 288us/step - loss: 0.5096 - accuracy: 0.8484
Epoch 4/10
60000/60000 [==============================] - 18s 295us/step - loss: 0.4035 - accuracy: 0.8808
Epoch 5/10
60000/60000 [==============================] - 17s 288us/step - loss: 0.3374 - accuracy: 0.9000
Epoch 6/10
60000/60000 [==============================] - 18s 304us/step - loss: 0.2893 - accuracy: 0.9155
Epoch 7/10
60000/60000 [==============================] - 18s 308us/step - loss: 0.2553 - accuracy: 0.9252
Epoch 8/10
60000/60000 [==============================] - 19s 311us/step - loss: 0.2305 - accuracy: 0.9323
Epoch 9/10
60000/60000 [==============================] - 18s 293us/step - loss: 0.2118 - accuracy: 0.9370
Epoch 10/10
60000/60000 [============

### 小結2:
* 在不調整其他參數下，僅比較RNN中SimpleRNN、GRU以及LSTM模型，LSTM與GRU的準確度皆較SimpleRNN好。
* LSTM與GRU引入了memory cell當作閘的概念，幫助前後神經網絡有記憶功能，較SimpleRNN更為強大。
* RNN模型應用在圖像識別(如本例)效果較CNN差，一般還是以CNN應用於圖像方面的問題較為頻繁。